pkg/arvo/lib/chat-json.hoon

```diff
@@ -0,0 +1,251 @@
+/-  *chat-store, *chat-view
+|%
+::
+++  slan  |=(mod/@tas |=(txt/@ta (need (slaw mod txt))))
+::
+++  seri                                              :::  serial
+  =,  dejs:format
+  ^-  $-(json serial)
+  (cu (slan %uv) so)
+::
+++  re                                                ::  recursive reparsers
+  |*  {gar/* sef/_|.(fist:dejs-soft:format)}
+  |=  jon/json
+  ^-  (unit _gar)
+  =-  ~!  gar  ~!  (need -)  -
+  ((sef) jon)
+::
+++  dank                                              ::  tank
+  ^-  $-(json (unit tank))
+  =,  ^?  dejs-soft:format
+  %+  re  *tank  |.  ~+
+  %-  of  :~
+    leaf+sa
+    palm+(ot style+(ot mid+sa cap+sa open+sa close+sa ~) lines+(ar dank) ~)
+    rose+(ot style+(ot mid+sa open+sa close+sa ~) lines+(ar dank) ~)
+  ==
+::
+++  eval                                              :::  %exp speech
+  :::  extract contents of an %exp speech, evaluating
+  :::  the {exp} if there is no {res} yet.
+  |=  a=json
+  ^-  [cord (list tank)]
+  =,  ^?  dejs-soft:format
+  =/  exp  ((ot expression+so ~) a)
+  %-  need
+  ?~  exp
+    [~ '' ~]
+  :+  ~  u.exp
+  =/  res  ((ot output+(ar dank) ~) a)
+  ?^  res
+    u.res
+  p:(mule |.([(sell (slap !>(..^zuse) (ream u.exp)))]~))  ::TODO  oldz
```

Comment:
This `+eval` currently allows scry, and should have a more useful subject -- see my comments in #1803

Reply:
This is a straight copy of %hall-json. I would certainly love it to be the best version of eval it could possibly be, but I don't understand how it works.

Reply:
Expanding this out a bit, which lines are particulatly confusing?
```
1  ::  u.exp=@t
2  =<  p  ^-  (each tang tang)
3  %-  mule  |.
4  =-  [-]~  :: tank -> (list tank)
5  %-  sell
6  %+  slap  !>(..zuse)
7  (ream u.exp)
```

And yes, to reiterate my comments on the "mule suddenly has scry by default" PR: if this used to be secure, and the exact same code is no longer secure, I am annoyed.

Reply:
I don't know what mule, sell, or slap do (apart from a vague hand-wavey "compile"),  or what the first argument to slap is in line 6.

Reply:
Read bottom-to-top, this is:

- parse `u.exp` to its ast (`+hoon`)
- compile that ast against the type of `..zuse` (standard library) and then run it against that actual value as well, producing a vase (pair of type and value)
- pretty-print that vase to a tank
- add a null so it's a singleton list of tanks (tang)
- run all of the above in a virtual environment so that if it crashes deterministically (syntax error in +ream, compile error in +slap, runtime error in +slap, prettyprint error in +sell) we get `[%| stack-trace=tang]` and if it succeeds we get `[%& result=tang]`
- get the tail of that value (which is a tang in both cases)

This gets you a tang of the result of the code or a stack trace of its failure.

Not that it matters, it's mostly fine to cargo cult this.

Reply:
It is *mostly* fine, expect for my repeated requests that you (or @Fang-) replace it with the implementation in the comments of #1803. This is a) to ensure a consistent subject for evals, and b) to prevent scry.

As @ohAitch notes, the issue of scry in `+mule` has been somewhat haphazardly handled, and these APIs need to be refactored. I haven't done this yet, so we need to jump through some extra hoops.

Reply:
@joemfb Would you be okay with this PR merging in the existing ++eval that Hall also uses, then in a subsequent PR merging in the new one?

Reply:
I guess, if there's some reason that the state-threading is just too onerous for now. At the very least, it should match the current hall-json:

https://github.com/urbit/urbit/blob/32052396d8a85ac7f2479e689dcce7ea2e266c9a/pkg/arvo/lib/hall-json.hoon#L575-L585

But the primary use-case of `+eval` in chat messages is sharing code snippets for support or pedagogy. It's not helpful in either situation for different clients to use different subjects.

Reply:
Ah, heads up, I just committed this to the chat-cli branch. 7f0142d

Reply:
Now using the eval function in the hall-json file you mentioned @joemfb. Thanks for posting that Mark. In your branch, we can include the updated eval functions for both JS chat client and chat-cli.